File: go/server/org_scoping_test.go
package server

import "testing"

// Org is the OUTER scope, applied before ownership.
//
// The gap this closes: Allows() returned true for ANY ownerless conversation
// (deliberate — th-909995 Option B keeps anonymous/emailless/legacy sessions
// reachable), and org was not consulted at all. So an ownerless conversation
// belonging to another org was readable by anyone holding its id — authorization
// resting on an unguessable UUID, which leaks through logs, referrers and
// screenshots.

func TestOrgIsCheckedBeforeOwnership(t *testing.T) {
	orgA := ConversationScope{Email: "a@example.com", OrgID: "org-a"}

	// THE GAP: an OWNERLESS conversation in another org used to be readable by
	// anyone, because the ownerless branch returned true before org was consulted.
	if orgA.Allows("", "org-b") {
		t.Error("an ownerless conversation in another org must NOT be visible")
	}
	// Same conversation in the caller's own org stays reachable — th-909995 intact.
	if !orgA.Allows("", "org-a") {
		t.Error("an ownerless conversation in the caller's own org must stay visible")
	}
	// An OWNED conversation in another org is denied even when the emails match,
	// so a shared email across orgs cannot cross the boundary.
	if orgA.Allows("a@example.com", "org-b") {
		t.Error("org must be checked before ownership, even for a matching email")
	}
	if !orgA.Allows("a@example.com", "org-a") {
		t.Error("owner + org both matching must be visible")
	}
}

func TestUnscopedStillSeesEverything(t *testing.T) {
	// Auth-disabled dev is the one unscoped path and must be untouched.
	unscoped := ConversationScope{Unscoped: true}
	for _, org := range []string{"", "org-a", "org-b"} {
		if !unscoped.Allows("someone@example.com", org) {
			t.Errorf("unscoped must see org %q", org)
		}
	}
}

func TestUnrecordedOrgFallsThroughToOwnership(t *testing.T) {
	// Rows created before org capture carry no org. Denying them would lock people
	// out of conversations they already own, so they fall through to the ownership
	// check rather than being denied outright.
	orgA := ConversationScope{Email: "a@example.com", OrgID: "org-a"}
	if !orgA.Allows("", "") {
		t.Error("an unrecorded-org ownerless conversation must stay reachable")
	}
	if !orgA.Allows("a@example.com", "") {
		t.Error("an unrecorded-org conversation the caller owns must stay reachable")
	}
	if orgA.Allows("b@example.com", "") {
		t.Error("ownership must still be enforced when the org is unrecorded")
	}
}

func TestCreatedConversationsRecordTheOrg(t *testing.T) {
	// End to end on the ACTUAL gap: an OWNERLESS conversation (an emailless
	// principal — the population th-909995 keeps reachable). Ownership cannot block
	// this one, so only the org check can. Use an owned conversation here and the
	// test passes with the org check removed, proving nothing.
	store := NewInMemorySessionStore()
	ctx := t.Context()

	a := ConversationScope{OrgID: "org-a"} // no Email → ownerless conversation
	session, err := store.CreateSession(ctx, "agent", "", "", a)
	if err != nil {
		t.Fatalf("create: %v", err)
	}
	if session.OwnerOrg != "org-a" || session.OwnerEmail != "" {
		t.Fatalf("session = %+v, want ownerless in org-a", session)
	}

	// Its own org can still resume it — th-909995 intact.
	if _, resumed, err := store.ResumeSession(ctx, "agent", "", "", a, session.ConversationID); err != nil || !resumed {
		t.Errorf("same org must resume its own ownerless conversation: resumed=%v err=%v", resumed, err)
	}

	// Another org must NOT, and must not be able to tell it from an unknown id.
	b := ConversationScope{OrgID: "org-b"}
	other, resumed, err := store.ResumeSession(ctx, "agent", "", "", b, session.ConversationID)
	if err != nil {
		t.Fatalf("cross-org resume: %v", err)
	}
	if resumed {
		t.Error("another org must never resume this ownerless conversation")
	}
	if other.ConversationID == session.ConversationID {
		t.Error("another org must not be bound to this conversation id")
	}
}

File: typescript/server/test/orgScoping.test.ts
/**
 * Org is the OUTER scope, applied before ownership.
 *
 * The gap this closes: `mayRead` returned true for ANY ownerless conversation
 * (deliberate — it keeps anonymous / emailless / legacy sessions reachable, see
 * PRs #308/#309) and never consulted org. So an ownerless conversation belonging
 * to another org was readable by anyone holding its id — authorization resting on
 * an unguessable UUID, which leaks through logs, referrers and screenshots.
 */
import { describe, expect, it } from 'vitest';

import { MockLlmProvider } from '@smooai/smooth-operator-core';

import type { AccessContext } from '../src/auth.js';
import { FrameDispatcher } from '../src/frameDispatcher.js';
import type { Frame } from '../src/protocol.js';
import { InMemorySessionStore } from '../src/sessionStore.js';

describe('conversation org scoping', () => {
    it('records the owning org at creation and never re-homes it on resume', async () => {
        const store = new InMemorySessionStore();
        // No userEmail → an OWNERLESS conversation: the case ownership cannot block,
        // so only the org check can. An owned conversation here would pass even with
        // the org check removed, proving nothing.
        const created = await store.createSession('agent', undefined, undefined, undefined, 'org-a');
        expect(created.orgId).toBe('org-a');
        expect(created.userEmail).toBeUndefined();

        const conv = await store.getConversation(created.conversationId);
        expect(conv?.orgId).toBe('org-a');

        // A resume from another org must not rewrite the conversation's org.
        const resumed = await store.createSession('agent', undefined, undefined, created.conversationId, 'org-b');
        expect(resumed.conversationId).toBe(created.conversationId);
        expect(resumed.orgId, 'a resume must inherit the original org, never the resumer’s').toBe('org-a');
    });

    it('leaves the org unrecorded when the caller has none', async () => {
        // Rows created before org capture carry no org; they must not be forced into
        // a bogus one, and the dispatcher falls through to ownership for them.
        const store = new InMemorySessionStore();
        const created = await store.createSession('agent', undefined, 'a@example.com');
        expect(created.orgId).toBeUndefined();
        expect((await store.getConversation(created.conversationId))?.orgId).toBeUndefined();
    });
});

/**
 * The gate itself, driven through the dispatcher — the store tests above only prove
 * the org is recorded, not that anything checks it.
 */
describe('the dispatcher gate checks org before ownership', () => {
    /** An authenticated principal in `org`, optionally with no email claim. */
    function principal(org: string, email?: string): AccessContext {
        return {
            principal: { sub: 'sub', org, role: 'basic', groups: [], ...(email ? { email } : {}) },
            isAnonymous: false,
            authEnabled: true,
        };
    }

    function connect(store: InMemorySessionStore, access: AccessContext) {
        const dispatcher = new FrameDispatcher({ store, chatClient: new MockLlmProvider(), access });
        const sink: Frame[] = [];
        return { sink, dispatch: (f: Record<string, unknown>) => dispatcher.dispatch(JSON.stringify(f), (x) => sink.push(x)) };
    }

    it('hides another org’s OWNERLESS conversation, which ownership alone cannot', async () => {
        const store = new InMemorySessionStore();

        // org-a, emailless → an ownerless conversation. Ownership can never block a
        // read of this, so the org check is the only thing that can.
        const a = connect(store, principal('org-a'));
        await a.dispatch({ type: 'action', action: 'create_conversation_session', requestId: 'r1', agentId: 'agent' });
        const conversationId = (a.sink[0]!.data as { conversationId: string }).conversationId;
        expect(conversationId).toBeTruthy();

        // org-b resuming that id must get a FRESH conversation, indistinguishable from
        // having passed an id that never existed.
        const b = connect(store, principal('org-b'));
        await b.dispatch({ type: 'action', action: 'create_conversation_session', requestId: 'r2', agentId: 'agent', conversationId });
        const bConversationId = (b.sink[0]!.data as { conversationId: string }).conversationId;
        expect(bConversationId, 'another org must not be bound to this conversation').not.toBe(conversationId);
    });

    it('lets the SAME org resume its own ownerless conversation', async () => {
        // The behaviour the ownerless rule exists to protect must survive.
        const store = new InMemorySessionStore();
        const a = connect(store, principal('org-a'));
        await a.dispatch({ type: 'action', action: 'create_conversation_session', requestId: 'r1', agentId: 'agent' });
        const conversationId = (a.sink[0]!.data as { conversationId: string }).conversationId;

        const again = connect(store, principal('org-a'));
        await again.dispatch({ type: 'action', action: 'create_conversation_session', requestId: 'r2', agentId: 'agent', conversationId });
        const resumedId = (again.sink[0]!.data as { conversationId: string }).conversationId;
        expect(resumedId, 'same org must keep its own conversation').toBe(conversationId);
    });
});
